File: python/day6.py
from utilities import read_input
from typing import List


class Group:
    def __init__(self, group: str):
        group = group.splitlines()
        self.num_of_people = len(group)
        self.answers = [0] * 26
        for person in group:
            for answer in person:
                self.answers[ord(answer) - 97] += 1

    def get_total_yes_answers(self):
        return sum([answer > 0 for answer in self.answers])

    def get_everyone_yes(self):
        return sum([answer == self.num_of_people for answer in self.answers])


def part_1(groups: List[Group]):
    return sum([group.get_total_yes_answers() for group in groups])


def part_2(groups: List[Group]):
    return sum([group.get_everyone_yes() for group in groups])


if __name__ == '__main__':
    print('Day 6')
    data = read_input('../inputs/day6.txt')
    groups = [Group(group) for group in data.split('\n\n')]
    print('\tPart 1: {}'.format(part_1(groups)))
    print('\tPart 2: {}'.format(part_2(groups)))

File: rust/src/day6.rs
use crate::utilities::read_block;

struct Group {
    num_of_people: usize,
    answers: Vec<usize>,
}

impl Group {
    fn new(group: String) -> Group {
        let group_answers: Vec<&str> = group.split('\n').collect();
        let num_of_people = group_answers.len();
        let mut answers = vec![0; 26];
        for person_answer in group_answers {
            for answer in person_answer.chars() {
                answers[answer as usize - 97] += 1;
            }
        }
        Group {
            num_of_people,
            answers,
        }
    }

    fn get_total_yes_answers(&self) -> usize {
        self.answers.iter().filter(|answer| **answer > 0).count()
    }

    fn get_everyone_yes(&self) -> usize {
        self.answers
            .iter()
            .filter(|answer| **answer == self.num_of_people)
            .count()
    }
}

fn part_1(groups: &Vec<Group>) {
    println!(
        "\tPart 1: {}",
        groups
            .iter()
            .map(|group| group.get_total_yes_answers())
            .sum::<usize>()
    );
}

fn part_2(groups: &Vec<Group>) {
    println!(
        "\tPart 2: {}",
        groups
            .iter()
            .map(|group| group.get_everyone_yes())
            .sum::<usize>()
    );
}

pub fn main() {
    println!("Day 6");
    let groups = read_block("../inputs/day6.txt")
        .into_iter()
        .map(Group::new)
        .collect();
    part_1(&groups);
    part_2(&groups);
}
